bgpd: VRF-Lite fix best path selection

iqras23 opened the pull request:
Description:
Incorrect behavior during best path selection for the imported routes.
Imported routes are always treated as eBGP routes.

Change is intended for fixing the issues related to
bgp best path selection for leaked routes:
- FRR does ecmp for the imported routes,
  even without any ecmp related config.
  If the same prefix is imported from two different VRFs,
  then we configure the route with ecmp even without
  any ecmp related config.
- Locally imported routes are preferred over imported
  eBGP routes.
  If there is a local route and eBGP learned route
  for the same prefix, if we import both the routes,
  imported local route is selected as best path.
- Same route is imported from multiple tenant VRFs,
  both imported routes point to the same VRF in nexthop.
- When the same route with same nexthop in two different VRFs
  is imported from those two VRFs, route is not installed as ecmp,
  even though we had ecmp config.

- During best path selection, while comparing the paths for imported routes,
  we should correctly refer to the original route i.e. the ultimate path.
- When the same route is imported from multiple VRF,
  use the correct VRF while installing in the FIB.
- When same route is imported from two different tenant VRFs,
  while comparing bgp path info as part of bgp best path selection,
  we should ideally also compare corresponding VRFs.

See-also: https://github.com/FRRouting/frr/files/7169555/FRR.and.Cisco.VRF-Lite.Behaviour.pdf

Co-authored-by: Santosh P K <sapk@vmware.com>
Co-authored-by: Kantesh Mundaragi <kmundaragi@vmware.com>
Signed-off-by: Iqra Siddiqui <imujeebsiddi@vmware.com>

donaldsharp reviewed (changes requested):
Let's break this commit up into it's component parts.  4 different fixes 4 different commits.

hsetnak commented:
> Let's break this commit up into it's component parts. 4 different fixes 4 different commits.

Hi Donald,
All 4 issues listed in the commit message are being fixed with only one single fix,
not different fixes here.
Basically we are fixing the best path selection for leaked routes.
May be we will reword the commit message.

Thank you!

pguibert6WIND commented:
Can you please illustrate with examples the commit log, please.



> Description:
> Change is intended for fixing the following issues related to vrf route leaking:
> 
>     1. FRR does ECMP for routes imported across VRF,
>        even without any ECMP related config.

The expectation is that by comparing path origin, the routes would not be considered as ECMP, right ?

> 
>     2. Imported routes are preferred over eBGP routes,
>        when a specific VRF is imported on both the eBGP peers.

I understand the importation process is based on an IGPpreference model. this is for avoiding preferring egp route over ibgp. right?

> 
>     3. While importing a specific prefix from multiple tenant VRFs into default VRF,
>        incorrect next-hop is picked for ECMP paths.
> 
>     4. ECMP doesn't work in default VRF,
>        when a route with same nexthop IP is imported from 2 tenant VRFs.
> 
what is difference with point 1? 



> 
> Co-authored-by: Kantesh Mundaragi [kmundaragi@vmware.com](mailto:kmundaragi@vmware.com)
> Signed-off-by: Iqra Siddiqui [imujeebsiddi@vmware.com](mailto:imujeebsiddi@vmware.com)

hsetnak commented:
> Can you please illustrate with examples the commit log, please.
> 
> > Description:
> > Change is intended for fixing the following issues related to vrf route leaking:
> > ```
> > 1. FRR does ECMP for routes imported across VRF,
> >    even without any ECMP related config.
> > ```
> 
> The expectation is that by comparing path origin, the routes would not be considered as ECMP, right ?
> 
> > ```
> > 2. Imported routes are preferred over eBGP routes,
> >    when a specific VRF is imported on both the eBGP peers.
> > ```
> 
> I understand the importation process is based on an IGPpreference model. this is for avoiding preferring egp route over ibgp. right?
> 
> > ```
> > 3. While importing a specific prefix from multiple tenant VRFs into default VRF,
> >    incorrect next-hop is picked for ECMP paths.
> > 
> > 4. ECMP doesn't work in default VRF,
> >    when a route with same nexthop IP is imported from 2 tenant VRFs.
> > ```
> 
> what is difference with point 1?
> 
> > Co-authored-by: Kantesh Mundaragi [kmundaragi@vmware.com](mailto:kmundaragi@vmware.com)
> > Signed-off-by: Iqra Siddiqui [imujeebsiddi@vmware.com](mailto:imujeebsiddi@vmware.com)

We have updated the commit message (elaborating further).

> Can you please illustrate with examples the commit log, please.
We have attached a document with all the testing scenarios, expected behavior and issues found.
If required we shall also provide examples.

> The expectation is that by comparing path origin, the routes would not be considered as ECMP, right ?
Correct. (Ref: bgp_path_info_cmp)

> I understand the importation process is based on an IGP preference model. this is for avoiding preferring egp route over ibgp. right?
Locally imported routes are preferred over imported eBGP routes.
If there is a local route and eBGP learned route for the same prefix,
if we import both the routes, imported local route is selected as best path.
One of the issue we are fixing.

> what is difference with point 1?
Same route is imported from multiple tenant VRFs,
both imported routes point to the same VRF in nexthop.

When the same route with same nexthop in two different VRFs
is imported from those two VRFs,
route is not installed as ecmp, even though we had ecmp config.

(Ref: bgp_path_info_nexthop_cmp)

pguibert6WIND commented on bgpd/bgp_routemap.c:
```diff
@@ -1148,7 +1148,7 @@ route_match_vrl_source_vrf(void *rule, const struct prefix *prefix,
 
-	if (path->extra == NULL)
+	if (path->extra == NULL || path->extra->bgp_orig == NULL)
 		return RMAP_NOMATCH;
```
this should be separate from rest of the code
it looks like a sanity code ?

hsetnak replied:
EVPN routes marked as imported routes, having bgp path info's extra,
where as they are not truly imported routes, so original bgp info will be NULL.

Thats the reason added the Null check for the original bgp info(bgp_path_info_nexthop_cmp bgp_mpath.c:L189).
Also added the same check here for route-maps.

We can always segregate this if required.

pguibert6WIND replied:
yes, please. add a specific commit for that, please

pguibert6WIND commented on bgpd/bgp_zebra.c:
```diff
@@ -1404,18 +1405,13 @@ void bgp_zebra_announce(struct bgp_dest *dest, const struct prefix *p,
-					nh_othervrf ? info->extra->bgp_orig
-						    : bgp,
+					nh_othervrf, bgp_orig,
 					&mpinfo_cp->attr->nexthop,
```
those changes in bgp_zebra should also be on separate commit. this is just code optimisation, here.right ?

hsetnak replied:
Yes, that is correct.
one extra check of "mpinfo->sub_type == BGP_ROUTE_IMPORTED"
and imo more readable.

Please let us know, we can always avoid the change in bgp_zebra totally.

hsetnak replied:
Ack, will fix it.

pguibert6WIND commented on bgpd/bgp_route.c:
```diff
@@ -583,6 +602,11 @@ static int bgp_path_info_cmp(struct bgp *bgp, struct bgp_path_info *new,
+	if (debug)
+		zlog_debug("%s: for bgp instance %s comparing path info",
+			   pfx_buf, bgp->name_pretty);
 
```
this debug code should be removed. no need to add it

hsetnak replied:
We were thinking this log would be useful,
in debugging bestpath selection of imported routes.

We will remove this log.

pguibert6WIND reviewed (changes requested):
apart from above request changes, the commit log is better, but needs some cleaning too:
- no need to write refer attach document if no pointer is given. perhaps See-also followed with an URL should be welcome.
- about sentence "- When routes are exported from VRF, a new bgp path info is created and
  the original path info is added as parent. Here new path info
  is created with type as BGP and subtype as imported.
  The new path will also have AS set to 0 with AS length as 1.
  During best path selection, since all the routes look the same,
  they are treated as ECMP and installed as ECMP."

what does it mean in relationship with code ?
Do you have apointer to share where the copy happens ? this could help it.

- I saw that the IGP metric has not been mentioned. I don't know if it in your perimeter too, but be aware that there are issues with IGP metric not propagated appropriately in imported VRF.

hsetnak commented:
> apart from above requested changes, the commit log is better, but needs some cleaning too:
> 
> * no need to write refer attach document if no pointer is given. perhaps See-also followed with an URL should be welcome.

We will fix it as See-also with an URL.

> * about sentence "- When routes are exported from VRF, a new bgp path info is created and
>   the original path info is added as parent. Here new path info
>   is created with type as BGP and subtype as imported.
>   The new path will also have AS set to 0 with AS length as 1.
>   During best path selection, since all the routes look the same,
>   they are treated as ECMP and installed as ECMP."
> 
> what does it mean in relationship with code ?
> Do you have a pointer to share where the copy happens ? this could help it.

Mentioned above to only highlight overall 'cause' of the issues.
https://gitlab.eng.vmware.com/core-build/nsx-frr/-/blob/nsbu-frr-stable/bgpd/bgp_mplsvpn.c#L608
Have not made any code changes there.

We shall remove the above sentence if it is not appropriate to the overall context of the commit log.

> * I saw that the IGP metric has not been mentioned. I don't know if it in your perimeter too, but be aware that there are issues with IGP metric not propagated appropriately in imported VRF.

If you could give some example or pointer,
we shall look into, if we can accommodate the changes to handle those.

pguibert6WIND commented:
> > apart from above requested changes, the commit log is better, but needs some cleaning too:
> > 
> > * no need to write refer attach document if no pointer is given. perhaps See-also followed with an URL should be welcome.
> 
> We will fix it as See-also with an URL.
> 
> > * about sentence "- When routes are exported from VRF, a new bgp path info is created and
> >   the original path info is added as parent. Here new path info
> >   is created with type as BGP and subtype as imported.
> >   The new path will also have AS set to 0 with AS length as 1.
> >   During best path selection, since all the routes look the same,
> >   they are treated as ECMP and installed as ECMP."
> > 
> > what does it mean in relationship with code ?
> > Do you have a pointer to share where the copy happens ? this could help it.
> 
> Mentioned above to only highlight overall 'cause' of the issues.
> https://gitlab.eng.vmware.com/core-build/nsx-frr/-/blob/nsbu-frr-stable/bgpd/bgp_mplsvpn.c#L608
> Have not made any code changes there.
> 
> We shall remove the above sentence if it is not appropriate to the overall context of the commit log.
> 
> > * I saw that the IGP metric has not been mentioned. I don't know if it in your perimeter too, but be aware that there are issues with IGP metric not propagated appropriately in imported VRF.
> 
> If you could give some example or pointer,
> we shall look into, if we can accommodate the changes to handle those.

Thanks for link pointer for reference where L3VPN code is performed.
for IGP calculation when importing metric, I can give you that link for EVPN:
https://github.com/FRRouting/frr/pull/9426

It is easy to have a similar setup with L3VPN.
with your setup, I think if you lean routes with an IGP like OSPF, and that one of the equipments redistributes routes with a metric higher, then you will have the problem:
eg:
interface eth0
  redistribute connected metric 20
exit

hsetnak commented on bgpd/bgp_mpath.c:
```diff
@@ -181,6 +181,20 @@ int bgp_path_info_nexthop_cmp(struct bgp_path_info *bpi1,
+	 * If both nexthops are same then check
+	 * if they belong to same VRF
+	 */
+	if (!compare && (bpi1->attr->nh_type != NEXTHOP_TYPE_BLACKHOLE)) {
```
Ack, will fix it.

hsetnak replied:
Ack, will fix it.

hsetnak commented on bgpd/bgp_route.c:
```diff
@@ -544,6 +544,25 @@ void bgp_path_info_path_with_addpath_rx_str(struct bgp_path_info *pi, char *buf,
+/*
+ * Get the ultimate path info.
+ */
+static struct bgp_path_info *bgp_get_bpi_ultimate(struct bgp_path_info *info)
```
Ack, will fix it.

hsetnak commented:
> > > apart from above requested changes, the commit log is better, but needs some cleaning too:
> > > 
> > > * no need to write refer attach document if no pointer is given. perhaps See-also followed with an URL should be welcome.
> > 
> > 
> > We will fix it as See-also with an URL.
> > > * about sentence "- When routes are exported from VRF, a new bgp path info is created and
> > >   the original path info is added as parent. Here new path info
> > >   is created with type as BGP and subtype as imported.
> > >   The new path will also have AS set to 0 with AS length as 1.
> > >   During best path selection, since all the routes look the same,
> > >   they are treated as ECMP and installed as ECMP."
> > > 
> > > what does it mean in relationship with code ?
> > > Do you have a pointer to share where the copy happens ? this could help it.
> > 
> > 
> > Mentioned above to only highlight overall 'cause' of the issues.
> > https://gitlab.eng.vmware.com/core-build/nsx-frr/-/blob/nsbu-frr-stable/bgpd/bgp_mplsvpn.c#L608
> > Have not made any code changes there.
> > We shall remove the above sentence if it is not appropriate to the overall context of the commit log.
> > > * I saw that the IGP metric has not been mentioned. I don't know if it in your perimeter too, but be aware that there are issues with IGP metric not propagated appropriately in imported VRF.
> > 
> > 
> > If you could give some example or pointer,
> > we shall look into, if we can accommodate the changes to handle those.
> 
> Thanks for link pointer for reference where L3VPN code is performed.
> for IGP calculation when importing metric, I can give you that link for EVPN:
> #9426
> 
> It is easy to have a similar setup with L3VPN.
> with your setup, I think if you lean routes with an IGP like OSPF, and that one of the equipments redistributes routes with a metric higher, then you will have the problem:
> eg:
> interface eth0
> redistribute connected metric 20
> exit

According to my understanding, below code should handle it, because metric check is past that:

L885:
	/* Here if these are imported routes then get ultimate pi for
	 * path compare. */
	new = bgp_get_bpi_ultimate(new);
	exist = bgp_get_bpi_ultimate(exist);
	newattr = new->attr;
	existattr = exist->attr;

Ref: 
https://github.com/FRRouting/frr/pull/9610/commits/9323a9aaa7074cbdc89c57d2e7b54f2a7fa46e70#diff-916a077bdc3f40b019663d23796308dd9ba7648f0a082bb975428b305c54b699R885
Metric check: https://github.com/FRRouting/frr/blob/master/bgpd/bgp_route.c#L1001

iqras23 commented:
ci:rerun

iqras23 commented:
Waiting for review.

iqras23 commented:
@here, 
Please remove labels libfrr, ospfv3, zebra from this PR.
They got added accidentally when we pushed some other changes over here by mistake, which we have fixed now..

ton31337 reviewed (approved):
LGTM

riw777 reviewed (approved):
looks good ... will wait for commit window and to see if anyone else wants to review

iqras23 commented:
ci:rerun

iqras23 commented:
ci:rerun

iqras23 commented:
ci:rerun

iqras23 commented:
ci:rerun

iqras23 commented:
ci:rerun

riw777 commented:
ci:rerun

iqras23 commented:
ci:rerun

iqras23 commented:
Requesting re-review.

iqras23 commented:
ci:rerun

iqras23 commented:
ci:rerun

iqras23 commented:
> @iqras23 -- any chance we can clear up the comments and get this committed? :-)

Hi All,

We have addressed all the review comments.
Also, we have made the below listed changes in the expected output for 
selected route in the script (test_bgp_l3vpn_to_bgp_vrf):
- With our changes for VRF-Lite: Fix best path selection,
   during best path selection, while comparing the paths for imported routes,
   we should correctly refer to the original route i.e. the ultimate path.
   In this case, when we have ibgp route and imported ibgp route for the same prefix,
   we do compare IGP metric which is same for both,
   So we proceed to comparing router-ids and selecting the best path.
- Corresponding changes for expected advertised route and the large community are also made.
Hence, requesting for review again.
Please review: @riw777 @donaldsharp @ton31337 @pguibert6WIND @louberger 

Thank you!

ton31337 commented on tests/topotests/bgp_vrf_lite_best_path_test/bgp_vrf_lite_best_path_topo2.json:
```diff
@@ -0,0 +1,1088 @@
+    "routers": {
+        "r1": {
+            "links": {
+                "r3-link1": {"ipv4": "13.1.1.1/24", "ipv6": "13::1:1/120", "vrf": "RED"},
```
Why do we use global IPv4 addresses for documentation? There are specific ranges for documentation: https://datatracker.ietf.org/doc/html/rfc5737 https://datatracker.ietf.org/doc/html/rfc3849

kuldeepkash replied:
> Why do we use global IPv4 addresses for documentation? There are specific ranges for documentation: https://datatracker.ietf.org/doc/html/rfc5737 https://datatracker.ietf.org/doc/html/rfc3849

HI @ton31337 , we are not using these IPs for documentation purpose, this is JSON file which is used for our topology.

ton31337 replied:
Topotest is a documentation let's say in this case. We should avoid using someone's IP ranges.

kuldeepkash replied:
> Topotest is a documentation let's say in this case. We should avoid using someone's IP ranges.

Hi @ton31337 , We have updated IPs accordingly. Thank you.

iqras23 commented:
ci:rerun

iqras23 commented:
ci:rerun

iqras23 commented:
Requesting for a re-review.

riw777 reviewed (approved):
looks good at this point, will wait for others to clear their comments

pguibert6WIND commented on bgpd/bgp_zebra.h:
```diff
@@ -23,6 +23,13 @@
 
+/* Macro to update bgp_original based on bpg_path_info */
+#define BGP_ORIGINAL_UPDATE(_bgp_orig, _mpinfo, _bgp)                          \
+	((_mpinfo->extra && _mpinfo->extra->bgp_orig                           \
```
can you add a protection with _mpinfo.
unless I misread, mpinfo can be NULL in evpn case. no?

hsetnak replied:
In case of evpn, what we observed is _mpinfo->extra->bgp_orig can be NULL.
Earlier we were only checking for _mpinfo->extra, where as in the case of evpn,
we had to add NULL check for original bgp path info.

EVPN routes marked as imported routes, having bgp path info's extra,
where as they are not truly imported routes, so original bgp info will be null (.
While comparing nexthop's vrf_id, we were dereferencing null.

Adding to that within bgp_zebra_announce() where we are using this macro,
we are already checking for NULL check with for loop.
`for (mpinfo = info; mpinfo; mpinfo = bgp_path_info_mpath_next(mpinfo)) {`
--------
Having said that, we can always add NULL check,
which is always better thing to do.

Please let us know.

pguibert6WIND reviewed (changes requested):
last change about value of mpimport.
other than that, lgtm.
